langchain4j-google-ai-gemini/src/test/java/dev/langchain4j/model/gemini/GeminiAiChatLanguageModelIT.java

```diff
@@ -0,0 +1,597 @@
+
+        // then
+        String jsonText = response.content().text();
+        assertThat(jsonText).contains("\"firstname\"");
+        assertThat(jsonText).contains("\"John\"");
+
+        assertThat(response.tokenUsage().inputTokenCount()).isEqualTo(25);
+        assertThat(response.tokenUsage().outputTokenCount()).isEqualTo(6);
+        assertThat(response.tokenUsage().totalTokenCount()).isEqualTo(31);
+    }
+
+    @Test
+    void should_support_multiturn_chatting() {
+        // given
+        GeminiAiChatLanguageModel gemini = GeminiAiChatLanguageModel.builder()
+            .apiKey(GEMINI_AI_KEY)
+            .modelName("gemini-1.5-flash")
+            .build();
+
+        List<ChatMessage> messages = new ArrayList<>();
+        messages.add(UserMessage.from("Hello, my name is Guillaume"));
+        messages.add(AiMessage.from("Hi, how can I assist you today?"));
+        messages.add(UserMessage.from("What is my name? Reply with just my name."));
+
+        // when
+        Response<AiMessage> response = gemini.generate(messages);
+
+        // then
+        assertThat(response.content().text()).contains("Guillaume");
+    }
+
+    @Test
+    void should_support_sending_images_as_base64() {
+        // given
+        GeminiAiChatLanguageModel gemini = GeminiAiChatLanguageModel.builder()
+            .apiKey(GEMINI_AI_KEY)
+            .modelName("gemini-1.5-flash")
+            .build();
+
+        // when
+        Response<AiMessage> response = gemini.generate(UserMessage.userMessage(
+//            ImageContent.fromToolExecReqToGFunCall(CAT_IMAGE_URL),
+            ImageContent.from(new String(Base64.getEncoder().encode(readBytes(CAT_IMAGE_URL))), "image/png"),
+            TextContent.from("Describe this image in a single word")));
+
+        // then
+        assertThat(response.content().text()).containsIgnoringCase("cat");
+    }
+
+    @Test
+    void should_support_sending_images_as_uri() {
+        // given
+        GeminiAiChatLanguageModel gemini = GeminiAiChatLanguageModel.builder()
+            .apiKey(GEMINI_AI_KEY)
+            .modelName("gemini-1.5-flash")
+            .build();
+
+        // when
+        Response<AiMessage> response = gemini.generate(UserMessage.userMessage(
+            ImageContent.from(CAT_IMAGE_URL), //TODO use file upload API
+            TextContent.from("Describe this image in a single word")));
+
+        // then
+        assertThat(response.content().text()).containsIgnoringCase("cat");
+    }
+
+    @Test
+    void should_support_text_file() {
+        // given
+        GeminiAiChatLanguageModel gemini = GeminiAiChatLanguageModel.builder()
+            .apiKey(GEMINI_AI_KEY)
+            .modelName("gemini-1.5-flash")
+            .build();
+
+        // when
+        Response<AiMessage> response = gemini.generate(UserMessage.userMessage(
+            TextFileContent.from(new String(Base64.getEncoder().encode(readBytes(MD_FILE_URL))), "text/markdown"),
```

Comment:
I had an impression that `TextFileContent` was introduced in order to specify file as a URL, but if I change it to `TextFileContent.from(MD_FILE_URL)`, it fails

Reply:
The variants of `*Content` classes don't yet fully work with just URLs.
Gemini has a file upload API that needs to be supported, to first upload files there, and then to point at the specific URL from that place. So external free-form URLs are not supported by Gemini out of the box.
But I'll try to see after the release what I can do here to improve this, to let users pass any URL and not just special URLs.